### PR TITLE
[2/2][Serve] Support dynamic log

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

- Follow up with the [pr](https://github.com/ray-project/ray/pull/40465) to implement the dynamic logging.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/application_state.py [hunk 1]
```diff
@@ -909,6 +911,11 @@ def build_serve_application(
         for deployment in deployments:
             deployment.set_options(version=code_version, _internal=True)
 
+            # If deployment logging config is not set and use application logging
+            # config if it is set.
+            if deployment._deployment_config.logging_config is None and logging_config:
+                deployment._deployment_config.logging_config = logging_config.dict()
```
Comment (edoakes) [inline python/ray/serve/_private/application_state.py:917]: this shouldn't use private APIs

File: python/ray/serve/_private/controller.py [hunk 2]
```diff
@@ -115,15 +117,33 @@ async def __init__(
...
+        # Try to read config from checkpoint
+        self.ray_worker_namespace = ray.get_runtime_context().namespace
+        self.controller_name = controller_name
+        self.gcs_client = GcsClient(address=ray.get_runtime_context().gcs_address)
+        kv_store_namespace = f"{self.controller_name}-{self.ray_worker_namespace}"
+        self.kv_store = RayInternalKVStore(kv_store_namespace, self.gcs_client)
+        log_config_checkpoint = self.kv_store.get(LOGGING_CONFIG_CHECKPOINT_KEY)
+        if log_config_checkpoint is not None:
+            logging_config = pickle.loads(log_config_checkpoint)
+        self.logging_config: LoggingConfig = logging_config
```
Comment (edoakes) [inline python/ray/serve/_private/controller.py:137]: Finding this logic hard to follow because the initial config and reconfigured configs are going down different paths (the former is an init arg, the latter goes through `reconfigure`) -- two possible suggestions for cleaning it up:

1. Don't pass the logging config as an init arg at all, and instead after creating the controller immediately call the `reconfigure_system_logging_config` method.
2. Add a helper method for getting the current logging config in the constructor. First check if there is a checkpointed logging config already. If so, use it. If not, call `reconfigure_system_logging_config` directly from the constructor.

I think either of these would be a good unification, lmk what you think.

File: python/ray/serve/_private/controller.py [hunk 3]
```diff
@@ -212,6 +228,22 @@ async def __init__(
...
+
+        self.long_poll_host.notify_changed(
+            LongPollNamespace.SYSTEM_LOGGING_CONFIG,
+            logging_config,
+        )
+        configure_component_logger(
+            component_name="controller",
+            component_id=str(os.getpid()),
+            logging_config=logging_config,
+        )
```
Comment (edoakes) [inline python/ray/serve/_private/controller.py:245]: put this in a shared method that's also called in constructor (to avoid duplicating the PID and "controller" constant)

File: python/ray/serve/_private/controller.py [hunk 4]
```diff
@@ -212,6 +228,22 @@ async def __init__(
...
+    def reconfigure_system_logging_config(self, logging_config: LoggingConfig):
+        if self.logging_config.version == logging_config.version:
+            return
+        self.kv.put(LOGGING_CONFIG_CHECKPOINT_KEY, pickle.dumps(logging_config))
+        self.logging_config = logging_config
+
+        self.long_poll_host.notify_changed(
+            LongPollNamespace.SYSTEM_LOGGING_CONFIG,
+            logging_config,
+        )
```
Comment (edoakes) [inline python/ray/serve/_private/controller.py:240]: I believe this also needs to be called when we load the logging config in the constructor. Another reason to unify the codepaths.

File: python/ray/serve/_private/controller.py (hunk 3 above)
Comment (sihanwang41) [reply python/ray/serve/_private/controller.py:241]: put configure logger inside the same function.

File: python/ray/serve/_private/controller.py (hunk 2 above)
Comment (sihanwang41) [reply python/ray/serve/_private/controller.py]: done!

File: python/ray/serve/_private/controller.py [hunk 5]
```diff
@@ -115,16 +117,30 @@ async def __init__(
...
+
+        self.long_poll_host = LongPollHost()
+        self.done_recovering_event = asyncio.Event()
+
+        # Try to read config from checkpoint
+        self.logging_config = None
+        log_config_checkpoint = self.kv_store.get(LOGGING_CONFIG_CHECKPOINT_KEY)
+        if log_config_checkpoint is not None:
+            logging_config = pickle.loads(log_config_checkpoint)
+        self.reconfigure_system_logging_config(logging_config)
```
Comment (edoakes) [inline python/ray/serve/_private/controller.py:142]: please add a comment about the precedence and why it's structured this way (i.e., that the one in the constructor is ignored if `reconfigure` is ever called)

**Review by edoakes — COMMENTED** [review]
Looks reasonable to me, LMK when it's fully ready for review w/ tests

**Comment by sihanwang41** [comment]
> Looks reasonable to me, LMK when it's fully ready for review w/ tests

hi @edoakes , added tests for config deploy and controller crash. PTAL!

**Comment by edoakes** [comment]
Sorry somehow I missed the ping. Will review in morning. You have a merge conflict FYI

**Comment by edoakes** [comment]
Update PR description please

File: python/ray/serve/_private/application_state.py [hunk 6]
```diff
@@ -1030,4 +1040,9 @@ def override_deployment_info(
         ):
             deployment.route_prefix = app_route_prefix
 
+    # When there is no deployment entries and app logging config is set,
+    # apply app logging config to all existing deployment logging config.
+    if not deployment_override_options and app_logging_config:
+        for deployment in list(deployment_infos.values()):
+            deployment.deployment_config.logging_config = app_logging_config
```
Comment (edoakes) [inline python/ray/serve/_private/application_state.py:1047]: I think this logic is not comprehensive. If there are `deployment_override_options`, but the list is not complete, then the deployments not in the list will not have their logging config overridden. Am I correct?

File: python/ray/serve/_private/config.py [hunk 7]
```diff
@@ -147,6 +155,20 @@ def user_config_json_serializable(cls, v):
...
+    def logging_config_valid(cls, v):
+        if v is None:
+            return v
+        if not isinstance(v, dict):
+            raise TypeError(
+                f"Got invalid type '{type(v)}' for logging_config. "
+                "Expected a dictionary."
+            )
+        if "log_level" not in v:
+            v["log_level"] = "INFO"
```
Comment (edoakes) [inline python/ray/serve/_private/config.py:168]: why do we need this instead of just setting the default in the schema?

File: python/ray/serve/_private/config.py [hunk 8]
```diff
@@ -147,6 +155,20 @@ def user_config_json_serializable(cls, v):
 
         return v
 
+    @validator("logging_config", always=True)
```
Comment (edoakes) [inline python/ray/serve/_private/config.py:158]: should we also be validating that the encoding is a known value?

File: python/ray/serve/_private/controller.py [hunk 9]
```diff
@@ -207,6 +218,29 @@ async def __init__(
...
+        )
+        configure_component_logger(
+            component_name="controller",
+            component_id=str(os.getpid()),
+            logging_config=logging_config,
+        )
+        logger.debug(
+            "Configure the serve controller logger "
+            f"with logging config: {self.logging_config}"
+        )
```
Comment (edoakes) [inline python/ray/serve/_private/controller.py:242]: Would make sure this only logs when the config is changed then make it info-level (it's useful)
```suggestion
        logger.debug(
            f"Serve controller logging config updated to: {self.logging_config}."
        )
```

File: python/ray/serve/_private/controller.py [hunk 10]
```diff
@@ -974,6 +1014,14 @@ def _save_cpu_profile_data(self) -> str:
                 "the RAY_SERVE_ENABLE_CPU_PROFILING env var."
             )
 
+    def _get_logger_state(self) -> Dict:
+        """Test purpose, get the log information."""
```
Comment (edoakes) [inline python/ray/serve/_private/controller.py:1018]: would like to avoid adding private methods for testing if possible, but if there's no other way it's OK
```suggestion
    def _get_logging_config(self) -> Dict:
        """Get the logging configuration (for testing purposes)."""
```
Would make this return a tuple instead of a dict with magic keys also

File: python/ray/serve/tests/test_api.py [hunk 11]
```diff
@@ -53,6 +67,157 @@ async def __call__(self):
         return {"count": self.count}
 
 
+class TestLoggingAPI:
```
Comment (edoakes) [inline python/ray/serve/tests/test_api.py:70]: These can go in `test_logging.py`

File: python/ray/serve/tests/test_api.py [hunk 12]
```diff
@@ -53,6 +67,157 @@ async def __call__(self):
...
+                else:
+                    assert len(logger.handlers) == 0
+                return
+
+        serve.run(Model.bind())
+
+        resp = requests.get("http://127.0.0.1:8000/")
+        assert resp.status_code == 200
+
+    def test_application_logging_overwrite(self, serve_instance):
```
Comment (edoakes) [inline python/ray/serve/tests/test_api.py:202]: please also add a test for overwriting per-deployment

File: python/ray/serve/schema.py [hunk 13]
```diff
@@ -69,6 +70,15 @@ class EncodingType(str, Enum):
     JSON = "JSON"
 
 
+@PublicAPI(stability="alpha")
+class AcessLoggingType(str, Enum):
```
Comment (zcin) [inline python/ray/serve/schema.py:74]:
```suggestion
class AccessLoggingType(str, Enum):
```

File: python/ray/serve/schema.py [hunk 14]
```diff
@@ -98,10 +108,11 @@ class Config:
...
         ),
     )
-    enable_access_log: bool = Field(
-        default=True,
+    access_log: Union[str, AcessLoggingType] = Field(
+        default="ALL",
         description=(
-            "Whether to enable access logs for each request. Default to True."
+            "Configure how to access the Serve logs. Default to 'ALL' to enable "
+            " streaming and file access logs together."
```
Comment (zcin) [inline python/ray/serve/schema.py:115]: Is this completely different from the `enable_access_log` field from before? I'm under the impression that "access logs" are the lines like
`(ServeReplica:default:Hello pid=15902) INFO 2023-11-07 13:10:11,907 Hello default#Hello#CvGNYA d7e93fde-3ec2-4172-8bbf-84377378a9d0 / default replica.py:730 - __CALL__ OK 92.1ms`

However it seems this new `access_log` field defines whether all Serve logs should be written to file and/or streamed to the driver.

File: python/ray/serve/schema.py [hunk 15]
```diff
@@ -116,17 +127,50 @@ def valid_encoding_format(cls, v):
 
         return v
 
+    @validator("access_log")
+    def valida_acccess_log(cls, v):
```
Comment (zcin) [inline python/ray/serve/schema.py:131]:
```suggestion
    def validate_acccess_log(cls, v):
```

File: python/ray/serve/_private/logging_utils.py [hunk 16]
```diff
@@ -198,37 +195,50 @@ def record_factory(*args, **kwargs):
...
+            os.path.join(logs_dir, log_file_name),
+            maxBytes=max_bytes,
+            backupCount=backup_count,
         )
-    else:
-        file_handler.setFormatter(ServeFormatter(component_name, component_id))
-    logger.addHandler(file_handler)
+        if RAY_SERVE_ENABLE_JSON_LOGGING:
+            logger.warning(
+                "'RAY_SERVE_ENABLE_JSON_LOGGING' is going to be deprecated, please use "
```
Comment (zcin) [inline python/ray/serve/_private/logging_utils.py:229]:
```suggestion
                "'RAY_SERVE_ENABLE_JSON_LOGGING' is deprecated, please use "
```

File: python/ray/serve/schema.py [hunk 17]
```diff
@@ -116,17 +127,50 @@ def valid_encoding_format(cls, v):
...
+        return v
+
+    def _should_enable_stream_logging(self):
+        return self.access_log in [AcessLoggingType.ALL, AcessLoggingType.STREAM_ONLY]
+
+    def _should_enable_file_logging(self):
+        return self.access_log in [AcessLoggingType.ALL, AcessLoggingType.FILE_ONLY]
+
+    @property
+    def version(self):
```
Comment (zcin) [inline python/ray/serve/schema.py:165]: Can we rename this and make it private? `version` seems a bit out of context (maybe just name it `_hash`?) if you don't know it's being used in `DeploymentVersion`, and this doesn't need to be exposed to the user.

File: python/ray/serve/_private/application_state.py [hunk 18]
```diff
@@ -875,6 +876,7 @@ def build_serve_application(
     code_version: str,
     name: str,
     args: Dict,
+    logging_config: Optional[LoggingConfig] = None,
```
Comment (zcin) [inline python/ray/serve/_private/application_state.py:879]: Can you update the docstring

File: python/ray/serve/_private/application_state.py [hunk 19]
```diff
@@ -908,6 +910,10 @@ def build_serve_application(
         deploy_args_list = []
         for deployment in deployments:
             is_ingress = deployment.name == ingress.name
+            # If app log config is set and deployment config is not set,
+            # set app config to deployment log config.
+            if deployment.logging_config is None and logging_config:
+                deployment.set_logging_config(logging_config.dict())
```
Comment (zcin) [inline python/ray/serve/_private/application_state.py:916]: Changing logging config doesn't change the code version right? If you set the deployment's logging config to the application logging config here, how can you distinguish in `override_deployment_info` between a logging config that was set at the deployment level versus a logging config that was inherited from the app level? After `build_serve_application` is run, seems like every deployment will have a logging config (so no overriding will happen in `override_deployment_info`?)

File: python/ray/serve/_private/api.py [hunk 20]
```diff
@@ -110,6 +111,7 @@ def _check_http_options(
 def _start_controller(
     http_options: Union[None, dict, HTTPOptions] = None,
     grpc_options: Union[None, dict, gRPCOptions] = None,
+    logging_config: Union[None, dict, LoggingConfig] = None,
```
Comment (zcin) [inline python/ray/serve/_private/api.py:114]: Can we change this to `system_logging_config` just like how we have `update_system_logging_config` in the client? It's a bit confusing to see `logging_config` in three different places

File: python/ray/serve/schema.py (hunk 17 above)
Comment (sihanwang41) [reply python/ray/serve/schema.py:165]: write __eq__ func

File: python/ray/serve/_private/application_state.py (hunk 6 above)
Comment (sihanwang41) [reply python/ray/serve/_private/application_state.py]: Updated logic:
1. if application logging config is set, apply to all existing deployment config.
2. if deployment logging config is set, will rely on existing for loop to update them.

File: python/ray/serve/_private/config.py (hunk 7 above)
Comment (sihanwang41) [reply python/ray/serve/_private/config.py:168]: the problem is that when are doing "unset=True" in https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/application_state.py#L956

Even though we have default value, we are not able to pass into the deployment status info. So I make it if no log_level is set, set it "INFO" to make sure we always pass the correct value down to the replica. Otherwise it will becomes empty string which is not valid for configuring the logging config.

File: python/ray/serve/_private/config.py (hunk 8 above)
Comment (sihanwang41) [reply python/ray/serve/_private/config.py:158]: Added. the DeploymentConfig is not user facing and got converted from Schema. So I am not too worried about the validation as long as we have strict check with the Schema.

File: python/ray/serve/tests/test_api.py (hunk 12 above)
Comment (sihanwang41) [reply python/ray/serve/tests/test_api.py]: done!

File: python/ray/serve/_private/application_state.py (hunk 19 above)
Comment (sihanwang41) [reply python/ray/serve/_private/application_state.py]: good catch, i removed the logic, relying on the override_deployment_info to do the trick.

File: python/ray/serve/_private/api.py (hunk 20 above)
Comment (sihanwang41) [reply python/ray/serve/_private/api.py]: done!

Comment (zcin) [reply python/ray/serve/_private/api.py]: Should we also change the config option in `ServeDeploySchema`? There's the same issue of the same option name being used in 3 places. Aware this is an API change, will leave it up to the team to decide.

File: python/ray/serve/_private/config.py (hunk 8 above)
Comment (sihanwang41) [reply python/ray/serve/_private/config.py:157]: I imported schema here to handle the default value and then do type-checking. So that we don't need to have other class which will be duplicated with LoggingConfig in the schema.
(e.g. AutoscalingConfig in config.py and schema.py)

File: python/ray/serve/schema.py (hunk 14 above)
Comment (sihanwang41) [reply python/ray/serve/schema.py]: sync offline, reverted back to the original api. We will add more functionality as customer asked for it.

**Comment by sihanwang41** [comment]
Hi @edoakes and @zcin , ptal, thx!

File: python/ray/serve/_private/api.py [hunk 21]
```diff
@@ -155,10 +157,16 @@ def _start_controller(
...
+    if system_logging_config is None:
+        system_logging_config = LoggingConfig()
+    elif isinstance(system_logging_config, dict):
+        system_logging_config = LoggingConfig(**system_logging_config)
+
     controller = ServeController.options(**controller_actor_options).remote(
         SERVE_CONTROLLER_NAME,
         http_config=http_options,
         grpc_options=grpc_options,
+        logging_config=system_logging_config,
```
Comment (edoakes) [inline python/ray/serve/_private/api.py:169]:
```suggestion
        system_logging_config=system_logging_config,
```

File: python/ray/serve/_private/application_state.py [hunk 22]
```diff
@@ -954,8 +957,16 @@ def override_deployment_info(
...
     config_dict = override_config.dict(exclude_unset=True)
+    app_logging_config = config_dict.get("logging_config", None)
     deployment_override_options = config_dict.get("deployments", [])
 
+    # If application loggin config is set, apply to all existing deployment config
+    if app_logging_config:
+        for _, info in deployment_infos.items():
+            info.deployment_config.logging_config = LoggingConfig(
+                **app_logging_config
+            ).dict()
```
Comment (edoakes) [inline python/ray/serve/_private/application_state.py:968]: Won't this override all deployments' logging config, even if there is one specified for the deployment? The desired behavior is being able to override deployment > application > system (this is standard practice, finest granularity should take precedence).

File: python/ray/serve/tests/test_logging.py [hunk 23]
```diff
@@ -280,6 +286,172 @@ def check_log_file(log_file: str, expected_regex: list):
...
+        serve.run(
+            Model2.bind(),
+            logging_config={"log_level": "DEBUG"},
+            name="app2",
+            route_prefix="/app2",
+        )
+        resp = requests.get("http://127.0.0.1:8000/app2").json()
+        # Make sure 'model_debug_level' log content does not exist
+        with pytest.raises(AssertionError):
+            check_log_file(resp["log_file"], [".*model_debug_level.*"])
```
Comment (edoakes) [inline python/ray/serve/tests/test_logging.py:452]: Should also assert that the INFO level logs exist, no?

File: python/ray/serve/_private/application_state.py (hunk 22 above)
Comment (edoakes) [reply python/ray/serve/_private/application_state.py:968]: Hmm from the tests it looks like what I'm saying is already the case. I'm not following how this works because it looks like you're overriding all deployments' logging config here with the application one.

**Review by edoakes — COMMENTED** [review]
Just one clarification then lgtm

File: python/ray/serve/_private/application_state.py (hunk 22 above)
Comment (sihanwang41) [reply python/ray/serve/_private/application_state.py:968]: Apply the application logging config here.
And then we have the for loop to override the deployment config if set. https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/application_state.py#L1020

So that deployment logging config will always be set lastly.

File: python/ray/serve/tests/test_logging.py (hunk 23 above)
Comment (sihanwang41) [reply python/ray/serve/tests/test_logging.py]: done.

File: python/ray/serve/_private/logging_utils.py [hunk 24]
```diff
@@ -203,32 +202,44 @@ def record_factory(*args, **kwargs):
     stream_handler.addFilter(log_to_stderr_filter)
     logger.addHandler(stream_handler)
 
-    logs_dir = get_serve_logs_dir()
-    os.makedirs(logs_dir, exist_ok=True)
-    if max_bytes is None:
+    if logging_config.enable_access_log:
```
Comment (zcin) [inline python/ray/serve/_private/logging_utils.py:205]: Seems like `enable_access_log` controls whether we write logs to log files under the directory `/tmp/ray/session_latest/logs/serve/`. However the description of `enable_access_log` in schema.py is "Whether to enable access logs for each request. Default to True."

**Review by zcin — COMMENTED** [review]
Some comments + 1 [pending comment](https://github.com/ray-project/ray/pull/40735/#discussion_r1387200882) from above, otherwise LGTM

File: python/ray/serve/_private/logging_utils.py (hunk 24 above)
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py:205]: Updated the description.

File: python/ray/serve/_private/application_state.py (hunk 22 above)
Comment (edoakes) [reply python/ray/serve/_private/application_state.py:968]: That was my confusion as well

Comment (sihanwang41) [reply python/ray/serve/_private/application_state.py:968]: synced with Cindy offline, we apply the application config out of the function. So that we don't need to track whether setting in the deployment code or yaml file.

Comment (edoakes) [reply python/ray/serve/_private/application_state.py]: Got it. So this works as-is? Can you leave a comment describing this btw it's pretty hard to see from the code

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/_private/application_state.py (hunk 22 above)
Comment (sihanwang41) [reply python/ray/serve/_private/application_state.py]: I need to write a unit test to cover this case. Yeah, I will leave comments too!

File: python/ray/serve/_private/application_state.py [hunk 25]
```diff
@@ -572,7 +573,18 @@ def _reconcile_target_deployments(self) -> None:
 
         # Set target state for each deployment
         for deployment_name, info in self._target_state.deployment_infos.items():
-            self.apply_deployment_info(deployment_name, info)
+            deploy_info = deepcopy(info)
```
Comment (zcin) [inline python/ray/serve/_private/application_state.py:576]: I think the logic itself looks good, just flagging that we will execute a deepcopy once per deployment for every control loop iteration. I think it would be more efficient to store a copy of this info object since it doesn't change between control loop iterations.

**Review by zcin — APPROVED** [review]

**Comment by sihanwang41** [comment]
@architkulkarni ready to merge, windows telemetry is flacky in master.
